File: experiments/10-runtimes/src/c/warm.c
#define _GNU_SOURCE

/* E2, partie 2 : trajectoire d'échauffement, référence compilée en C. À chaque pas, une
 * chaîne de hachage puis un produit scalaire strict sur n éléments, chronométrés un par un
 * dès le premier pas du processus. Mêmes boucles que E1. */

#include <inttypes.h>
#include <sched.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define NOINLINE __attribute__((noinline))
#define K UINT64_C(0x9e3779b97f4a7c15)

static NOINLINE uint64_t hash_chain(const uint64_t *u, size_t n, uint64_t h) {
    for (size_t i = 0; i < n; ++i) h = (h ^ u[i]) * K;
    return h;
}

static NOINLINE double dot(const double *a, const double *b, size_t n) {
    double s = 0.0;
    for (size_t i = 0; i < n; ++i) s += a[i] * b[i];
    return s;
}

/* CPU autorisés (Cpus_allowed_list), virgules remplacées par « ; » pour le CSV. */
static void allowed_cpus(char *text, size_t size) {
    char line[256];
    snprintf(text, size, "inconnu");
    FILE *f = fopen("/proc/self/status", "r");
    while (f && fgets(line, sizeof line, f))
        if (sscanf(line, "Cpus_allowed_list: %63s", text) == 1) break;
    if (f) fclose(f);
    for (char *c = text; *c; ++c) if (*c == ',') *c = ';';
}

static uint64_t monotonic_ns(void) {
    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    return (uint64_t)t.tv_sec * UINT64_C(1000000000) + (uint64_t)t.tv_nsec;
}

int main(int argc, char **argv) {
    const char *input = NULL, *output = NULL, *meta = NULL, *mode = "c", *generate = NULL;
    unsigned long run_id = 0, steps = 300, n_log2 = 16, count = 0, seed = 20260923;
    for (int i = 1; i + 1 < argc; i += 2) {
        const char *o = argv[i], *v = argv[i + 1];
        if (!strcmp(o, "--input")) input = v;
        else if (!strcmp(o, "--output")) output = v;
        else if (!strcmp(o, "--meta")) meta = v;
        else if (!strcmp(o, "--mode")) mode = v;
        else if (!strcmp(o, "--generate")) generate = v;
        else if (!strcmp(o, "--run-id")) run_id = strtoul(v, NULL, 10);
        else if (!strcmp(o, "--steps")) steps = strtoul(v, NULL, 10);
        else if (!strcmp(o, "--n-log2")) n_log2 = strtoul(v, NULL, 10);
        else if (!strcmp(o, "--count")) count = strtoul(v, NULL, 10);
        else if (!strcmp(o, "--seed")) seed = strtoul(v, NULL, 10);
    }
    if (generate) {   /* données communes : count entiers de 64 bits, SplitMix64 */
        FILE *f = fopen(generate, "wb");
        uint64_t state = seed;
        for (unsigned long i = 0; f && i < count; ++i) {
            uint64_t z = (state += K);
            z = (z ^ (z >> 30)) * UINT64_C(0xbf58476d1ce4e5b9);
            z = (z ^ (z >> 27)) * UINT64_C(0x94d049bb133111eb);
            z ^= z >> 31;
            if (fwrite(&z, sizeof z, 1, f) != 1) return EXIT_FAILURE;
        }
        return f && fclose(f) == 0 ? EXIT_SUCCESS : EXIT_FAILURE;
    }
    if (!input || !output || !meta || run_id == 0 || n_log2 > 22) {
        fprintf(stderr, "usage : warm --input F --output CSV --meta CSV --run-id N --mode M [--steps S] [--n-log2 L]\n");
        return EXIT_FAILURE;
    }
    size_t n = (size_t)1 << n_log2;
    uint64_t *u = malloc(2 * n * sizeof *u);
    double *a = malloc(n * sizeof *a), *b = malloc(n * sizeof *b);
    uint64_t *elapsed = malloc(2 * steps * sizeof *elapsed), *result = malloc(2 * steps * sizeof *result);
    int *cpus = malloc(2 * steps * sizeof *cpus);
    FILE *f = fopen(input, "rb");
    if (!u || !a || !b || !elapsed || !result || !cpus || !f || fread(u, sizeof *u, 2 * n, f) != 2 * n) {
        fprintf(stderr, "lecture impossible\n");
        return EXIT_FAILURE;
    }
    fclose(f);
    for (size_t i = 0; i < n; ++i) {
        a[i] = (double)(u[i] >> 11) * 0x1.0p-53;
        b[i] = (double)(u[n + i] >> 11) * 0x1.0p-53;
    }
    for (unsigned long s = 0; s < steps; ++s) {
        uint64_t t0 = monotonic_ns();
        result[2 * s] = hash_chain(u, n, 0);
        uint64_t t1 = monotonic_ns();
        double d = dot(a, b, n);
        uint64_t t2 = monotonic_ns();
        memcpy(&result[2 * s + 1], &d, sizeof d);
        elapsed[2 * s] = t1 - t0;
        elapsed[2 * s + 1] = t2 - t1;
        cpus[2 * s] = cpus[2 * s + 1] = sched_getcpu();
    }
    FILE *out = fopen(output, "w");
    if (!out) return EXIT_FAILURE;
    fprintf(out, "run_id,mode,step,workload,elements,elapsed_ns,result,cpu\n");
    for (unsigned long s = 0; s < 2 * steps; ++s)
        fprintf(out, "%lu,%s,%lu,%s,%zu,%" PRIu64 ",%016" PRIx64 ",%d\n", run_id, mode, s / 2,
                s % 2 ? "dot" : "hash_chain", n, elapsed[s], result[s], cpus[s]);
    FILE *m = fopen(meta, "w");
    if (!m || fclose(out) != 0) return EXIT_FAILURE;
    char cpus_allowed[64];
    allowed_cpus(cpus_allowed, sizeof cpus_allowed);
    fprintf(m, "run_id,mode,runtime,jit,cpus_allowed\n%lu,%s,gcc %s,none,%s\n", run_id, mode, __VERSION__, cpus_allowed);
    return fclose(m) == 0 ? EXIT_SUCCESS : EXIT_FAILURE;
}

File: experiments/10-runtimes/src/java/Warm.java
// E2, partie 2 : trajectoire d'échauffement en Java (HotSpot). Mêmes boucles et même
// protocole que src/c/warm.c : à chaque pas, chaîne de hachage puis produit scalaire strict.

import java.io.IOException;
import java.io.PrintWriter;
import java.lang.management.ManagementFactory;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;

public final class Warm {
    static final long K = 0x9e3779b97f4a7c15L;

    static long hashChain(long[] u, int n, long h) {
        for (int i = 0; i < n; i++) {
            h = (h ^ u[i]) * K;
        }
        return h;
    }

    static double dot(double[] a, double[] b, int n) {
        double s = 0.0;
        for (int i = 0; i < n; i++) {
            s += a[i] * b[i];
        }
        return s;
    }

    static String allowedCpus() throws IOException {
        for (String line : Files.readAllLines(Path.of("/proc/self/status"))) {
            if (line.startsWith("Cpus_allowed_list:")) {
                return line.substring(line.indexOf(':') + 1).trim().replace(',', ';');
            }
        }
        return "inconnu";
    }

    public static void main(String[] args) throws IOException {
        HashMap<String, String> opt = new HashMap<>();
        for (int i = 0; i + 1 < args.length; i += 2) {
            opt.put(args[i], args[i + 1]);
        }
        int n = 1 << Integer.parseInt(opt.getOrDefault("--n-log2", "16"));
        int steps = Integer.parseInt(opt.getOrDefault("--steps", "300"));
        String mode = opt.get("--mode");
        long runId = Long.parseLong(opt.get("--run-id"));

        ByteBuffer raw = ByteBuffer.wrap(Files.readAllBytes(Path.of(opt.get("--input")))).order(ByteOrder.LITTLE_ENDIAN);
        long[] u = new long[2 * n];
        raw.asLongBuffer().get(u, 0, 2 * n);
        double[] a = new double[n];
        double[] b = new double[n];
        for (int i = 0; i < n; i++) {
            a[i] = (double) (u[i] >>> 11) * 0x1.0p-53;
            b[i] = (double) (u[n + i] >>> 11) * 0x1.0p-53;
        }
        long[] elapsed = new long[2 * steps];
        long[] result = new long[2 * steps];
        for (int s = 0; s < steps; s++) {
            long t0 = System.nanoTime();
            result[2 * s] = hashChain(u, n, 0);
            long t1 = System.nanoTime();
            double d = dot(a, b, n);
            long t2 = System.nanoTime();
            result[2 * s + 1] = Double.doubleToRawLongBits(d);
            elapsed[2 * s] = t1 - t0;
            elapsed[2 * s + 1] = t2 - t1;
        }
        try (PrintWriter out = new PrintWriter(opt.get("--output"))) {
            out.println("run_id,mode,step,workload,elements,elapsed_ns,result,cpu");
            for (int s = 0; s < 2 * steps; s++) {
                out.printf("%d,%s,%d,%s,%d,%d,%016x,-1%n", runId, mode, s / 2, s % 2 == 1 ? "dot" : "hash_chain",
                        n, elapsed[s], result[s]);
            }
        }
        String arguments = String.join(" ", ManagementFactory.getRuntimeMXBean().getInputArguments());
        try (PrintWriter meta = new PrintWriter(opt.get("--meta"))) {
            meta.println("run_id,mode,runtime,jit,cpus_allowed");
            meta.printf("%d,%s,java %s,%s,%s%n", runId, mode, System.getProperty("java.vm.version"),
                    arguments.contains("-Xint") ? "disabled" : "enabled", allowedCpus());
        }
    }
}
